### PR TITLE
[SPARK-7176] [ml] Add validation functionality to Param

### PR DESCRIPTION
**Pull request by jkbradley**
Main change: Added isValid field to Param.  Modified all usages to use isValid when relevant.  Added helper methods in ParamValidate.

Also overrode Params.validate() in:
- CrossValidator + model
- Pipeline + model

I made a few updates for the elastic net patch:
- I changed "tol" to "convergenceTol"
- I added some documentation

This PR is Scala + Java only.  Python will be in a follow-up PR.

CC: @mengxr

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala [hunk 1]
```diff
@@ -305,6 +439,8 @@ private[spark] object Params {
   }
 }
 
+abstract class JavaParams extends Params
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: I'll add doc here

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala [hunk 2]
```diff
@@ -86,6 +86,13 @@ class Pipeline extends Estimator[PipelineModel] {
   def setStages(value: Array[PipelineStage]): this.type = { set(stages, value); this }
   def getStages: Array[PipelineStage] = getOrDefault(stages)
 
+  override def validate(paramMap: ParamMap): Unit = {
+    val map = extractParamMap(paramMap)
+    getStages.foreach {
+      case pStage: Params => pStage.validate(map)
+    }
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: Should add `case _ =>`.

File: mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala [hunk 3]
```diff
@@ -32,10 +32,14 @@ import org.apache.spark.sql.types.DataType
...
 
   /**
-   * number of features
+   * Number of features.  Should be > 0.
+   * (default = 2^18^)
    * @group param
    */
-  val numFeatures = new IntParam(this, "numFeatures", "number of features")
+  val numFeatures = new IntParam(this, "numFeatures", "number of features (> 0)",
+    ParamValidate.gt[Int](0))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala]: `[Int]` is not necessary.

File: mllib/src/main/scala/org/apache/spark/ml/feature/PolynomialExpansion.scala [hunk 4]
```diff
@@ -38,9 +38,12 @@ class PolynomialExpansion extends UnaryTransformer[Vector, Vector, PolynomialExp
 
   /**
    * The polynomial degree to expand, which should be larger than 1.
+   * Default: 2
    * @group param
    */
-  val degree = new IntParam(this, "degree", "the polynomial degree to expand")
+  val degree = new IntParam(this, "degree", "the polynomial degree to expand",
+    ParamValidate.gt[Int](2))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/PolynomialExpansion.scala]: `gt(1)`. Users may set degree to `1` to turn off polynomial expansion in tuning.

File: mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala [hunk 5]
```diff
@@ -93,6 +95,13 @@ class CrossValidator extends Estimator[CrossValidatorModel] with CrossValidatorP
   /** @group setParam */
   def setNumFolds(value: Int): this.type = set(numFolds, value)
 
+  override def validate(paramMap: ParamMap): Unit = {
+    val map = extractParamMap(paramMap)
+    getEstimatorParamMaps.foreach { eMap =>
+      getEstimator.validate(map ++ eMap)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala]: Should be `getEstimator.validate(eMap ++ paramMap)`. Estimator doesn't need CV params.

File: project/SparkBuild.scala [hunk 6]
```diff
@@ -467,7 +467,7 @@ object Unidoc {
         "mllib.evaluation", "mllib.feature", "mllib.random", "mllib.stat.correlation",
         "mllib.stat.test", "mllib.tree.impl", "mllib.tree.loss",
         "ml", "ml.attribute", "ml.classification", "ml.evaluation", "ml.feature", "ml.param",
-        "ml.tuning"
+        "ml.recommendation", "ml.regression", "ml.tuning"
```
Comment (mengxr) [inline project/SparkBuild.scala]: This should be in a different PR.

File: python/pyspark/mllib/linalg.py [hunk 7]
```diff
@@ -39,7 +39,8 @@
     IntegerType, ByteType
 
 
-__all__ = ['Vector', 'DenseVector', 'SparseVector', 'Vectors', 'DenseMatrix', 'Matrices']
+__all__ = ['Vector', 'DenseVector', 'SparseVector', 'Vectors',
+           'DenseMatrix', 'Matrices', 'SparseMatrix']
```
Comment (mengxr) [inline python/pyspark/mllib/linalg.py]: This should be in a different PR.

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala [hunk 8]
```diff
@@ -65,38 +90,129 @@ class Param[T] (val parent: Params, val name: String, val doc: String) extends S
   }
 }
 
+/**
+ * Factory methods for common validation functions for [[Param.isValid]].
+ * The numerical methods only support Int, Long, Float, and Double.
+ */
+object ParamValidate {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: We might need `sealed abstract class ParamValidate` to make it show up correctly in Java. For the name, I recommend `ParamValidators` instead of `ParamValidate`.

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: Hm, it was working before...I'll fix it and also check out the Java docs

**Comment by jkbradley** [comment]
Btw, I made a few updates for the elastic net patch:
* I changed "tol" to "convergenceTol"
* I added some documentation

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala [hunk 9]
```diff
@@ -89,16 +94,19 @@ private[recommendation] trait ALSParams extends Params with HasMaxIter with HasR
   def getImplicitPrefs: Boolean = getOrDefault(implicitPrefs)
 
   /**
-   * Param for the alpha parameter in the implicit preference formulation.
+   * Param for the alpha parameter in the implicit preference formulation (>= 0).
+   * Default: 1.0
    * @group param
    */
-  val alpha = new DoubleParam(this, "alpha", "alpha for implicit preference")
+  val alpha = new DoubleParam(this, "alpha", "alpha for implicit preference",
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: @mengxr  Please confirm this isValid is correct

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala [hunk 10]
```diff
@@ -136,7 +147,7 @@ private[recommendation] trait ALSParams extends Params with HasMaxIter with HasR
 
   setDefault(rank -> 10, maxIter -> 10, regParam -> 0.1, numUserBlocks -> 10, numItemBlocks -> 10,
     implicitPrefs -> false, alpha -> 1.0, userCol -> "user", itemCol -> "item",
-    ratingCol -> "rating", nonnegative -> false)
+    ratingCol -> "rating", nonnegative -> false, checkpointInterval -> 10)
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: @mengxr  Defaults for ALS differ between ml and mllib.  Please confirm values in ml.

**Comment by jkbradley** [comment]
@mengxr  Hopefully that fixed everything.  I'm sending separate PRs for those other 2 small, unrelated changes

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala (hunk 9 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: Yes.

File: mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala (hunk 10 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala]: Yes, I changed it to 10 in spark.ml. We may checkpoint too often with 3.

**Comment by mengxr** [comment]
@jkbradley Let's try to keep this PR minimal. We discussed `tol` vs. `convergenceTol` in the elastic net PR and `tol` seems to be sufficient, which is also used in scikit-learn and MATLAB as the param name for convergence tolerance.

**Comment by jkbradley** [comment]
@mengxr  I'm fine taking tol out of this PR.  I missed the discussion on the elastic net PR.  Using "tol" goes against usual Spark naming conventions, which aim to make names easily understood.  I'll remove it though.

**Comment by jkbradley** [comment]
updated

**Comment by mengxr** [comment]
@jkbradley Could you merge the master branch? The changes to LinearRegression documentation may conflict with #5767. If there are documentation updates, we can do that in a separate PR.

**Comment by jkbradley** [comment]
Hopefully that did it.

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!
